// n8n/dist/config/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.schema = void 0;
const config_1 = require("@n8n/config");
const di_1 = require("@n8n/di");
exports.schema = {
    executions: {
        mode: {
            doc: 'If it should run executions directly or via queue',
            format: ['regular', 'queue'],
            default: 'regular',
            env: 'EXECUTIONS_MODE',
        },
        timeout: {
            doc: 'Max run time (seconds) before stopping the workflow execution',
            format: Number,
            default: -1,
            env: 'EXECUTIONS_TIMEOUT',
        },
        maxTimeout: {
            doc: 'Max execution time (seconds) that can be set for a workflow individually',
            format: Number,
            default: 3600,
            env: 'EXECUTIONS_TIMEOUT_MAX',
        },
        saveDataOnError: {
            doc: 'What workflow execution data to save on error',
            format: ['all', 'none'],
            default: 'all',
            env: 'EXECUTIONS_DATA_SAVE_ON_ERROR',
        },
        saveDataOnSuccess: {
            doc: 'What workflow execution data to save on success',
            format: ['all', 'none'],
            default: 'all',
            env: 'EXECUTIONS_DATA_SAVE_ON_SUCCESS',
        },
        saveExecutionProgress: {
            doc: 'Whether or not to save progress for each node executed',
            format: Boolean,
            default: false,
            env: 'EXECUTIONS_DATA_SAVE_ON_PROGRESS',
        },
        saveDataManualExecutions: {
            doc: 'Save data of executions when started manually via editor',
            format: Boolean,
            default: true,
            env: 'EXECUTIONS_DATA_SAVE_MANUAL_EXECUTIONS',
        },
    },
    userManagement: {
        isInstanceOwnerSetUp: {
            doc: "Whether the instance owner's account has been set up",
            format: Boolean,
            default: false,
        },
        authenticationMethod: {
            doc: 'How to authenticate users (e.g. "email", "ldap", "saml")',
            format: ['email', 'ldap', 'saml'],
            default: 'email',
        },
    },
    endpoints: {
        rest: {
            format: String,
            default: di_1.Container.get(config_1.GlobalConfig).endpoints.rest,
        },
    },
    ai: {
        enabled: {
            format: Boolean,
            default: di_1.Container.get(config_1.GlobalConfig).ai.enabled,
        },
    },
};
